thetis/diagnostics.py

```diff
@@ -141,9 +153,42 @@ def __init__(self, field_2d, hessian_2d, gradient_2d=None, **kwargs):
         kwargs.setdefault('solver_parameters', sp)
         self.solver = LinearVariationalSolver(prob, **kwargs)
 
-    @PETSc.Log.EventDecorator("thetis.HessianRecoverer2D.solve")
-    def solve(self):
+    @PETSc.Log.EventDecorator("thetis.HessianRecoverer2D.calculate")
+    def calculate(self):
         self.solver.solve()
         self.hessian_2d.assign(self._hessian)
         if self.gradient_2d is not None:
             self.gradient_2d.assign(self._gradient)
+
+
+class KineticEnergyCalculator(DiagnosticCalculator):
+    r"""
+    Class for calculating dynamic pressure (i.e. kinetic energy),
+
+    .. math::
+        E_K = \frac12 \rho \|\mathbf{u}\|^2,
+
+    where :math:`\rho` is the water density and :math:`\mathbf{u}`
+    is the velocity.
+    """
+    density = FiredrakeScalarExpression(
+        physical_constants['rho0'], help='Fluid density').tag(config=True)
+
+    @unfrozen
+    @PETSc.Log.EventDecorator("thetis.KineticEnergyCalculator.__init__")
+    def __init__(self, uv, ke, density=None, horizontal=False):
+        """
+        :arg uv: scalar expression for the fluid velocity.
+        :arg ke: :class:`Function` to hold calculated kinetic energy.
+        :kwarg density: fluid density.
+        :kwarg horizontal: consider the horizontal velocity only?
+        """
+        if density is not None:
+            self.density = density
+        self.ke = ke
+        u_sq = uv[0]*uv[0] + uv[1]*uv[1] if horizontal else dot(uv, uv)
+        self.ke_expr = 0.5*self.density*u_sq
+
+    @PETSc.Log.EventDecorator("thetis.KineticEnergyCalculator.calculate")
+    def calculate(self):
+        self.ke.interpolate(self.ke_expr)
```

Comment:
Should we construct a reusable `Interpolator` object? Are there cases where the used would actually like to project instead of interpolate? Projection preserves total energy I think. Maybe we could add a kwarg to choose.

Reply:
Ah yes, good idea!